scripts/train/train.py

```diff
@@ -164,170 +165,315 @@ def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
+                                         'load_weights_only',
+                                         must_exist=False,
+                                         default_value=False)
+    load_ignore_keys: Optional[List[str]] = pop_config(cfg,
+                                                       'load_ignore_keys',
+                                                       must_exist=False,
+                                                       default_value=None)
+    # Enable autoresume from model checkpoints if possible
+    autoresume_default = not save_overwrite and not save_weights_only
+    autoresume = pop_config(cfg,
+                            'autoresume',
+                            must_exist=False,
+                            default_value=autoresume_default)
+
+    # Pop known unused parameters.
+    pop_config(cfg, 'data_local', must_exist=False)
+    pop_config(cfg, 'data_remote', must_exist=False)
+    pop_config(cfg, 'global_seed', must_exist=False)
+    pop_config(cfg, 'global_train_batch_size', must_exist=False)
+    pop_config(cfg, 'n_gpus', must_exist=False)
+    pop_config(cfg, 'device_train_grad_accum', must_exist=False)
+
+    # Warn users for unused parameters
+    for key in cfg:
+        warnings.warn(
+            f'Unused parameter {key} found in cfg. Please check your yaml to ensure this parameter is necessary.'
+        )
+
+    # Warn if fsdp is enabled but user only has 1 GPU
     if dist.get_world_size() == 1 and fsdp_config is not None:
         warnings.warn(
             'FSDP is not applicable for single-GPU training. Reverting to DDP.')
-        cfg.pop('fsdp_config')
         fsdp_config = None
 
-    init_context = process_init_device(cfg.model, fsdp_config)
+    # Initialize context
+    init_context = process_init_device(model_config, fsdp_config)
+    if fsdp_config:
+        log_config(om.create(fsdp_config))
 
-    # build tokenizer
-    tokenizer = build_tokenizer(cfg.tokenizer)
+    # Build tokenizer
+    tokenizer = build_tokenizer(tokenizer_config)
 
     # Build Model
     print('Initializing model...')
     with init_context:
-        if cfg.get('lora',
-                   None) is not None:  # frozen model + trainable lora modules
+        if lora_config is not None:  # frozen model + trainable lora modules
             model: ComposerHFCausalLM = build_composer_peft_model(
-                cfg.model, cfg.lora, tokenizer)
+                model_config, lora_config, tokenizer)
             print_trainable_parameters(model)  # should not be 100%
         else:  # standard model
-            model = build_composer_model(cfg.model, tokenizer)
-    cfg.n_params = sum(p.numel() for p in model.parameters())
-    print(f'{cfg.n_params=:.2e}')
+            model = build_composer_model(model_config, tokenizer)
+
+    # Log number of parameters
+    n_params = sum(p.numel() for p in model.parameters())
+    log_config(om.create({'n_params': n_params}))
```

Comment:
same, this needs to happen after trainer init

Reply:
yep yep